In [101]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [102]:
df=pd.read_csv('C:/Users/Lenovo/Downloads/data.csv')

In [114]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 4600 entries, 0 to 4599
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   price          4600 non-null   float64
 1   bedrooms       4600 non-null   float64
 2   bathrooms      4600 non-null   float64
 3   sqft_living    4600 non-null   int64  
 4   sqft_lot       4600 non-null   int64  
 5   floors         4600 non-null   float64
 6   waterfront     4600 non-null   int64  
 7   view           4600 non-null   int64  
 8   condition      4600 non-null   int64  
 9   sqft_above     4600 non-null   int64  
 10  sqft_basement  4600 non-null   int64  
 11  yr_built       4600 non-null   int64  
 12  yr_renovated   4600 non-null   int64  
 13  year           4600 non-null   int32  
 14  month          4600 non-null   int32  
 15  day            4600 non-null   int32  
 16  city_freq      4600 non-null   float64
 17  statezip_freq  4600 non-null   float64
dtypes: float64(6), int3

In [116]:
df.shape

(4600, 18)

In [105]:
#fixing date
df['date']=pd.to_datetime(df['date'])
year=df['date'].dt.year
month=df['date'].dt.month
day=df['date'].dt.day
df=df.drop('date',axis=1)

df['year']=year
df['month']=month
df['day']=day


In [106]:
#drop street
df['street'].unique()
df=df.drop('street',axis=1)

In [107]:
df['city'].value_counts()

city
Seattle                1573
Renton                  293
Bellevue                286
Redmond                 235
Kirkland                187
Issaquah                187
Kent                    185
Auburn                  176
Sammamish               175
Federal Way             148
Shoreline               123
Woodinville             115
Maple Valley             96
Mercer Island            86
Burien                   74
Snoqualmie               71
Kenmore                  66
Des Moines               58
North Bend               50
Covington                43
Duvall                   42
Lake Forest Park         36
Bothell                  33
Newcastle                33
Tukwila                  29
Vashon                   29
SeaTac                   29
Enumclaw                 28
Carnation                22
Normandy Park            18
Fall City                11
Clyde Hill               11
Medina                   11
Black Diamond             9
Ravensdale                7
Pacific        

In [108]:
#city
df['city_freq']=df['city'].map(df['city'].value_counts(normalize=True))
df=df.drop('city', axis=1)


In [109]:
print(df['statezip'].nunique())
df['statezip_freq']=df['statezip'].map(df['statezip'].value_counts(normalize=True))
df=df.drop('statezip',axis=1)

77


In [110]:
#country
print(df['country'].nunique())
df=df.drop('country', axis=1)


1


In [111]:
X=df.drop("price", axis=1)
y=df['price']

In [112]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test=train_test_split(X, y, test_size=0.2, random_state=42)

In [113]:
#step 1
from sklearn.linear_model import LinearRegression
model=LinearRegression()
#step 2
model.fit(X_train, y_train)
predictions=model.predict(X_test)

In [123]:
print(np.sort(model.coef_))



[-5.53315911e+04 -1.39503344e+03 -5.46362381e+02 -3.13057931e-01
  6.54836185e-11  9.30321826e+00  6.57735122e+01  1.20594304e+02
  1.86367816e+02  3.03992616e+03  4.00759573e+03  1.78436402e+04
  2.79553692e+04  3.77112938e+04  4.31928970e+04  4.29010926e+05
  4.35163212e+05]


In [131]:
coef_df=pd.DataFrame({
    "feature":['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors',
       'waterfront', 'view', 'condition', 'sqft_above', 'sqft_basement',
       'yr_built', 'yr_renovated', 'year', 'month', 'day', 'city_freq',
       'statezip_freq'],
    "coefficient": model.coef_
})
coef_df['abs_coef']=coef_df['coefficient'].abs()
coef_df=coef_df.sort_values('abs_coef',ascending=False)

In [132]:
print(coef_df)

          feature   coefficient      abs_coef
16  statezip_freq  4.351632e+05  4.351632e+05
15      city_freq  4.290109e+05  4.290109e+05
0        bedrooms -5.533159e+04  5.533159e+04
14            day  4.319290e+04  4.319290e+04
13          month  3.771129e+04  3.771129e+04
12           year  2.795537e+04  2.795537e+04
11   yr_renovated  1.784364e+04  1.784364e+04
10       yr_built  4.007596e+03  4.007596e+03
9   sqft_basement  3.039926e+03  3.039926e+03
1       bathrooms -1.395033e+03  1.395033e+03
2     sqft_living -5.463624e+02  5.463624e+02
8      sqft_above  1.863678e+02  1.863678e+02
7       condition  1.205943e+02  1.205943e+02
6            view  6.577351e+01  6.577351e+01
5      waterfront  9.303218e+00  9.303218e+00
3        sqft_lot -3.130579e-01  3.130579e-01
4          floors  6.548362e-11  6.548362e-11


so the feature that has the strongest effect here is statezip_freq that i transformed from string to a frequency float that calculate the frequency of appearing across the other features

In [ ]:
y_mean=y_test.mean()
y_base=np.full_like(y_test, y_mean)

In [134]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
print("MAE", mean_absolute_error(y_test,predictions))
print( "RMSE", np.sqrt(mean_squared_error(y_test, predictions)))
print("R2", r2_score(y_test, predictions))

MAE 204851.0620752982
RMSE 990887.5542759225
R2 0.03724893974203658


In [138]:
baseline_rmse= print("baseline RSMA", np.sqrt(mean_squared_error(y_test, y_base)))

baseline RSMA 1009874.4179782724


it looks like having much large error number for each metric ( that the prediction is far away from the real value )
and the model didn't learn much from features need to work more on data preprocessing 
the model rsma is close to the baseline rsma wich is the average so that tells about the model it is week and can't predict right values its all around the average
for R2 its really close to 0 and it explains 3.7% of the variance